File: Swift/ExampleTests/common/ShoppingCartString.swift
public struct ShoppingCartString: CustomStringConvertible, Equatable {
    public let items: String
    public let shipping: String
    public let discount: String
    public let totalPrice: String
    public let shippingAddress: AddressString
    public let billingAddress: AddressString

    public init(items: String, shipping: String, discount: String, totalPrice: String, shippingAddress: AddressString, billingAddress: AddressString) {
        self.items = items
        self.shipping = shipping
        self.discount = discount
        self.totalPrice = totalPrice
        self.shippingAddress = shippingAddress
        self.billingAddress = billingAddress
    }

    public init(fromArray v: [String]) {
        self.items = v.count > 0 ? v[0] : ""
        self.shipping = v.count > 1 ? v[1] : ""
        self.discount = v.count > 2 ? v[2] : ""
        self.totalPrice = v.count > 3 ? v[3] : ""
        self.shippingAddress = AddressString(fromArray: [])
        self.billingAddress = AddressString(fromArray: [])
    }

    /// Builds from the text form, e.g. Money as "25 USD".
    public static func fromText(_ text: String) -> ShoppingCartString {
        let parts = Tokens.require(text, 6, "ShoppingCart")
        return ShoppingCartString(items: parts[0], shipping: parts[1], discount: parts[2], totalPrice: parts[3], shippingAddress: AddressString.fromText(parts[4]), billingAddress: AddressString.fromText(parts[5]))
    }

    public var description: String {
        return Tokens.token(items) + " " + Tokens.token(shipping) + " " + Tokens.token(discount) + " " + Tokens.token(totalPrice) + " " + Tokens.nested(shippingAddress.description) + " " + Tokens.nested(billingAddress.description)
    }

    public static let dncString = "?DNC?"

    public static func == (a: ShoppingCartString, b: ShoppingCartString) -> Bool {
        return (a.items == dncString || b.items == dncString || a.items == b.items)
            && (a.shipping == dncString || b.shipping == dncString || a.shipping == b.shipping)
            && (a.discount == dncString || b.discount == dncString || a.discount == b.discount)
            && (a.totalPrice == dncString || b.totalPrice == dncString || a.totalPrice == b.totalPrice)
            && a.shippingAddress == b.shippingAddress
            && a.billingAddress == b.billingAddress
    }
}

File: Python/ExampleTests/common/shopping_cart_string.py
from .address_string import AddressString
from .address_string import AddressString
from . import tokens
DNC_STRING = '?DNC?'


class ShoppingCartString:
    def __init__(self, items: str = '', shipping: str = '', discount: str = '', total_price: str = '', shipping_address: 'AddressString' = None, billing_address: 'AddressString' = None):
        self.items = items
        self.shipping = shipping
        self.discount = discount
        self.total_price = total_price
        self.shipping_address = shipping_address
        self.billing_address = billing_address

    @classmethod
    def from_list(cls, values):
        v = list(values)
        return cls(
            v[0] if len(v) > 0 else '',
            v[1] if len(v) > 1 else '',
            v[2] if len(v) > 2 else '',
            v[3] if len(v) > 3 else '',
            v[4] if len(v) > 4 else '',
            v[5] if len(v) > 5 else ''
        )

    @classmethod
    def from_text(cls, text):
        """Builds from the text form, e.g. Money as '25 USD'."""
        parts = tokens.require(text, 6, 'ShoppingCart')
        return cls(
            parts[0],
            parts[1],
            parts[2],
            parts[3],
            AddressString.from_text(parts[4]),
            AddressString.from_text(parts[5])
        )

    def __str__(self):
        return (tokens.token(self.items) + ' ' +
                tokens.token(self.shipping) + ' ' +
                tokens.token(self.discount) + ' ' +
                tokens.token(self.total_price) + ' ' +
                tokens.nested(str(self.shipping_address)) + ' ' +
                tokens.nested(str(self.billing_address)))

    def _key(self):
        return (self.items, self.shipping, self.discount, self.total_price, self.shipping_address, self.billing_address)

    def __eq__(self, other):
        if not isinstance(other, ShoppingCartString):
            return NotImplemented
        return all(a == b or a == DNC_STRING or b == DNC_STRING
                   for a, b in zip(self._key(), other._key()))

    def __hash__(self):
        return hash(self._key())

    def __repr__(self):
        return f'ShoppingCartString({self})'
